# Source detection with EpiLearn

The `Detection` task classifies every node of a graph from its observed state -- for
example "was this the node where the outbreak started?". It is a node classification
problem, so unlike `Forecast` it is not naturally a rolling-window problem: you choose
the splits yourself and call `train_model`.

In 0.1.0 `train_model` no longer splits the data for you. It takes three split dicts
built by `Dataset.generate_dataset(...)`, and `evaluate_model` takes a split dict too
(see [MIGRATION.md](../MIGRATION.md)).

In [1]:
import os
import torch
import matplotlib.pyplot as plt

# The dataset paths below are relative to the repository root, so step out of examples/.
# (Guarded so that re-running the cell is harmless.)
if os.path.basename(os.getcwd()) == "examples":
    os.chdir("..")

from epilearn.models.Spatial.GCN import GCN
from epilearn.models.Spatial.SAGE import SAGE
from epilearn.models.Spatial.GAT import GAT
from epilearn.models.Spatial.GIN import GIN

from epilearn.data import Dataset          # was UniversalDataset before 0.1.0
from epilearn.utils import utils, transforms
from epilearn.utils import simulation
from epilearn.tasks.detection import Detection

/home/ec2-user/miniconda3/envs/epilearn/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


### Configs

In [2]:
# initial settings
device = 'cpu'
torch.manual_seed(7)

lookback = 1 # inputs size
horizon = 2  # predicts size; also seen as number of classes

epochs = 20 # training epochs
batch_size = 25 # training batch size

### Initialize Dataset

In [3]:
# load toy dataset
dataset = Dataset()
dataset.load_toy_dataset()

# the detection task needs class labels per node; turn the counts into a binary label
dataset.y = (dataset.y > dataset.y.median()).long()
print(dataset, "| labels:", dataset.y.unique().tolist())

Dataset(Spatiotemporal, x=(539, 47, 4), y=(539, 47), graph=(47, 47)) | labels: [0, 1]


### Initialize model and task

We use `GCN` here. The task derives the model's constructor arguments from the data and
passes `num_nodes`, `num_features` and the window sizes to the prototype, so a prototype
has to tolerate those keywords: `GCN` does (it accepts `**kwargs`), while `GAT`, `SAGE`
and `GIN` have stricter signatures and need a thin wrapper class.

In [4]:
task = Detection(prototype=GCN, dataset=dataset, lookback=lookback, horizon=horizon, device=device)

### Add transformations

In [5]:
# transformation = transforms.Compose({
#                                  'features':[transforms.normalize_feat()],
#                                  'graph': [transforms.normalize_adj()],
#                                  'dynamic_graph': [transforms.normalize_adj()],
#                                  'states': []
#                                  })
transformation = transforms.Compose({
                                 'features':[],
                                 'graph': [],
                                 })
# apply_now=True normalizes the data in place, because we build the splits by hand below
dataset.set_transforms(transformation, apply_now=True)

Dataset(Spatiotemporal, x=(539, 47, 4), y=(539, 47), graph=(47, 47))

### Build the splits

`generate_dataset` turns a time range into sliding windows and returns a **dict**
(`features`, `targets`, `states`, `graph`, `dynamic_graph`). Remember to pass
`adj=dataset.graph`: without it `split['graph']` is silently `None` and a graph model has
nothing to convolve over.

In [6]:
def make_split(start, end):
    return dataset.generate_dataset(X=dataset.x[start:end], Y=dataset.y[start:end],
                                    adj=dataset.graph,
                                    lookback_window_size=lookback, horizon_size=1)

train_split, val_split, test_split = make_split(0, 300), make_split(300, 400), make_split(400, 539)

for name, split in [('train', train_split), ('val', val_split), ('test', test_split)]:
    print(f"{name:5}: features={tuple(split['features'].shape)} "
          f"targets={tuple(split['targets'].shape)} graph={tuple(split['graph'].shape)}")

train: features=(299, 1, 47, 4) targets=(299, 47, 1) graph=(47, 47)
val  : features=(99, 1, 47, 4) targets=(99, 47, 1) graph=(47, 47)
test : features=(138, 1, 47, 4) targets=(138, 47, 1) graph=(47, 47)


### Train model

In [7]:
result = task.train_model(train_split=train_split,
                          val_split=val_split,
                          test_split=test_split,
                          train_loss='ce',
                          val_loss='ce',
                          epochs=epochs,
                          batch_size=batch_size)
print("test loss:", result['loss'])


Final Training loss: 0.2611139652629693
Final Validation loss: 0.2609338164329529

Predicting Progress...
test loss: 0.25155359506607056


### Evaluate model

`evaluate_model` now takes the split you want to score. For detection it reports
classification metrics plus bootstrap confidence intervals.

In [8]:
evaluation = task.evaluate_model(dataset=test_split)
print("accuracy:", evaluation['accuracy'], "| macro F1:", evaluation['macro_f1'])


Predicting Progress...

--- Bootstrap Confidence Intervals (n=100) ---
Computing bootstrap estimates...
Accuracy 95% CI:  [0.8890, 0.9026]
F1 Score 95% CI:  [0.8138, 0.8356]

COMPREHENSIVE MODEL EVALUATION - SOURCE DETECTION

--- Classification Metrics ---
Accuracy:         0.8964
Macro Precision:  0.8895
Macro Recall:     0.7896
Macro F1:         0.8257
Mean Confidence:  0.9212

--- Per-Class Metrics ---
Class 0 - Precision: 0.8798, Recall: 0.6019, F1: 0.7148
Class 1 - Precision: 0.8993, Recall: 0.9774, F1: 0.9367

accuracy: 0.8963922262191772 | macro F1: 0.8257350623607635


### Train on a simulated dataset

Detection is most natural on simulated outbreaks, where we know who patient zero was.
`NetSIR` propagates an SIR process over a graph; we record the final state of every node
as the input and the initial state as the label.

In [9]:
# Simulation Process
from epilearn.models.SpatialTemporal.NetworkSIR import NetSIR

# generate 100 samples
num_nodes = 25
# generate random static graph: 25 nodes
initial_graph = simulation.get_random_graph(num_nodes=num_nodes, connect_prob=0.15)
initial_states = torch.zeros(num_nodes, 3) # [S,I,R]
initial_states[:, 0] = 1

x = []
y = []
for i in range(100):
    # set infected individual
    states = initial_states.clone()
    idx = torch.randint(0, num_nodes, (1,))
    states[idx.item(), 0] = 0
    states[idx.item(), 1] = 1

    model = NetSIR(num_nodes=initial_graph.shape[0], horizon=100,
                   infection_rate=0.01, recovery_rate=0.0384) # infection_rate, recover_rate, fixed_population
    preds = model(states, initial_graph, steps=None)
    x.append(torch.nn.functional.one_hot(preds[-1].argmax(1), num_classes=3))
    y.append(states.argmax(1))

x = torch.stack(x).float()
y = torch.stack(y).long()
print("x:", tuple(x.shape), "(samples, nodes, states) | y:", tuple(y.shape))

x: (100, 25, 3) (samples, nodes, states) | y: (100, 25)


In [10]:
dataset = Dataset(x=x, y=y, graph=initial_graph)
dataset.set_transforms(transformation, apply_now=True)

train_split, val_split, test_split = make_split(0, 60), make_split(60, 80), make_split(80, 100)
print("train features:", tuple(train_split['features'].shape),
      "| targets:", tuple(train_split['targets'].shape))

task = Detection(prototype=GCN, dataset=dataset, lookback=lookback, horizon=horizon, device=device)

train features: (59, 1, 25, 3) | targets: (59, 25, 1)


In [11]:
result = task.train_model(train_split=train_split,
                          val_split=val_split,
                          test_split=test_split,
                          train_loss='ce',
                          val_loss='ce',
                          epochs=epochs,
                          batch_size=5)

evaluation = task.evaluate_model(dataset=test_split, compute_bootstrap_ci=False)
print("accuracy:", evaluation['accuracy'], "| macro F1:", evaluation['macro_f1'])


Final Training loss: 0.17245866979161897
Final Validation loss: 0.16890133917331696

Predicting Progress...

Predicting Progress...

COMPREHENSIVE MODEL EVALUATION - SOURCE DETECTION

--- Classification Metrics ---
Accuracy:         0.9600
Macro Precision:  0.4800
Macro Recall:     0.5000
Macro F1:         0.4898
Mean Confidence:  0.9613

--- Per-Class Metrics ---
Class 0 - Precision: 0.9600, Recall: 1.0000, F1: 0.9796
Class 1 - Precision: 0.0000, Recall: 0.0000, F1: 0.0000

accuracy: 0.9599999785423279 | macro F1: 0.48979589343070984


Do not read too much into the accuracy here: only 1 of 25 nodes is the source, so a model
that predicts "not the source" everywhere already scores 96%. Watch the macro F1 and the
per-class recall instead, and note that this setup only shows the model the *final* state
of the outbreak.

### Next steps

`tutorial.ipynb` Part 6 builds a custom spatiotemporal detector that sees a *window* of
states after the outbreak instead of only the final one, which is a much harder and more
realistic version of this problem.